src/System.Diagnostics.Process/src/System/Diagnostics/ProcessWaitState.Unix.cs

```diff
@@ -511,5 +489,78 @@ private Task WaitForExitAsync(CancellationToken cancellationToken = default(Canc
             });
         }
 
+        private bool TryReapChild()
+        {
+            lock (_gate)
+            {
+                if (_exited)
+                {
+                    return false;
+                }
+
+                // Try to get the state of the child process
+                int status;
+                int waitResult = Interop.Sys.WaitPid(_processId, out status, Interop.Sys.WaitPidOptions.WNOHANG);
+
+                if (waitResult == _processId)
+                {
+                    // Process has exited
+                    if (Interop.Sys.WIfExited(status))
+                    {
+                        _exitCode = Interop.Sys.WExitStatus(status);
+                    }
+                    else if (Interop.Sys.WIfSignaled(status))
+                    {
+                        const int ExitCodeSignalOffset = 128;
+                        _exitCode = ExitCodeSignalOffset + Interop.Sys.WTermSig(status);
+                    }
+
+                    SetExited();
+                    return true;
+                }
+                else if (waitResult == 0)
+                {
+                    // Process is still running
+                    return false;
+                }
+                else if (waitResult == -1)
+                {
+                    Debug.Fail("Unexpected errno value from waitpid");
```

Comment:
ECHILD is impossible?

Reply:
If someone else has `waitpid` our child (which they shouldn't), we'd get `ECHILD`. The code then does what is expected: `SetExited`.
For non-child processes (which were previously detected as ECHILD), the detection is now in `CheckForNonChildExit`.